add support for self-signed certificates

little-dude opened the pull request:
This is untested, and based on my PR for hyper-native-tls (https://github.com/sfackler/hyper-native-tls/pull/7), but I'd like to know if you think this is the right direction.

This is a step toward fixing https://github.com/seanmonstar/reqwest/issues/15

little-dude commented:
`hyper-native-tls` now supports custom certificates.
I also updated the PR to also support hostname verification to be disabled.
It compiles, but I'll test this tonight for real.

little-dude commented:
travis build is fixed, but appveyor seems to be stuck or pending.

seanmonstar commented on Cargo.toml:
```diff
@@ -11,7 +11,7 @@ categories = ["web-programming::http-client"]
 [dependencies]
 hyper = "0.10.2"
-hyper-native-tls = "0.2"
+hyper-native-tls = { git = "https://github.com/sfackler/hyper-native-tls" }
```
Just making a note here for myself to not merge until we can remove the git dependency.

seanmonstar commented on src/client.rs:
```diff
@@ -47,21 +49,123 @@ pub struct Client {
+    }
+
+    /// Disable hostname verification
+    pub fn disable_hostname_verification(&mut self) {
```
Probably should have a scary name like the others, `danger_disable_hostname_verification`.

seanmonstar commented on src/client.rs:
```diff
@@ -47,21 +49,123 @@ pub struct Client {
+        let mut tls_client = try_!(
+            self.inner.build().map_err(|e| ::hyper::Error::Ssl(Box::new(e))));
+        if ! self.hostname_verification {
+            tls_client.dannger_disable_hostname_verification(true);
```
`dannger`, with 2 `n`s?

little-dude replied:
[Yes](https://github.com/sfackler/hyper-native-tls/blob/master/src/lib.rs#L142) probably a typo, I'll make a PR to fix this.

little-dude replied:
It's been fixed already. I'll update the PR.

seanmonstar commented on src/client.rs:
```diff
@@ -47,21 +49,123 @@ pub struct Client {
-        let mut client = try_!(new_hyper_client());
-        client.set_redirect_policy(::hyper::client::RedirectPolicy::FollowNone);
-        Ok(Client {
+impl From<NativeTlsClient> for Client {
```
If we did the conservative change like I suggest in the review comment, we'd probablt want to remove this impl.

little-dude commented:
Hi @seanmonstar !

Thanks for the kind words :)

I'm not sure to understand you comment.
You'd prefer not to have a dependency to `hyper_native_tls`?
Or you just don't want to expose it (which means we need to re-export `hyper_native_tls::Certificate` as I did in 9b4edca)?

seanmonstar commented:
reqwest already depends on hyper-native-tls, so I'm not suggesting we remove that. I'm saying let's see if we can prevent exposing that details as part of the public API. There is an interest in trying to get reqwest to a stable 1.0, so part of that requires not exporting or showing any types from other crates that aren't 1.0 stable yet.

So, to a user, all they see is a `reqwest::ClientBuilder`, and a `reqwest::Certificate`. As far as they know, it could be any TLS library underneath there. Sort of like with `libflate`: the user doesn't have any way of accidentally or purposefully touching something in the reqwest API that exposes that `libflate` is the gzip implementation used internally. Did I make that more clear?

little-dude commented:
I think I got it. Do you thinkg the last two commits address this?

seanmonstar commented on src/client.rs:
````diff
@@ -47,21 +49,134 @@ pub struct Client {
+    /// let _ = ::std::fs::File::open("my-cert.der").unwrap().read_to_end(&mut buf).unwrap();
+    /// let cert = reqwest::Certificate::from_der(&buf).unwrap();
+    /// # }
+    /// ```
````
The Rust community is starting to encourage a different way of writing doc examples, that don't use `unwrap` and stuff. It could be written like this:

```
/// ```
/// # use std::fs::File;
/// # fn cert() -> Result<(), Box<std::error::Error>> {
/// let mut buf = Vec::new();
/// File::open("my_cert.der")?
///     .read_to_end(&mut buf)?;
/// let cert = reqwest::Certificate::from_der(&buf)?;
/// # drop(cert);
/// # Ok(())
/// # }
/// ```
```

seanmonstar commented on src/client.rs:
````diff
@@ -47,21 +49,134 @@ pub struct Client {
+///
+///     // [...]
+/// }
+/// ```
````
Similar comment about the example here: put it some sort of `# fn foo() -> Result<(), Box<std::error::Error>>`, and use `?` instead of `unwrap`.

seanmonstar commented on src/client.rs:
````diff
@@ -47,21 +49,134 @@ pub struct Client {
+/// }
+/// ```
+///
+pub struct TlsClient {
````
I actually liked that naming you have before, `ClientBuilder`. We can eventually move several of the config properties to the builder and remove the need for the `Mutex` and atomics...

seanmonstar commented on src/client.rs:
```diff
@@ -47,21 +49,134 @@ pub struct Client {
+    }
+
+    /// Returns a `Client` that uses this `TlsClient` configuration.
+    pub fn into_client(self) -> ::Result<Client> {
```
Prefer the method name `build`.

seanmonstar commented on src/client.rs:
```diff
@@ -47,21 +49,134 @@ pub struct Client {
 
+    /// Add a custom root certificate. This can be used to connect to a server that has a
+    /// self-signed certificate for example.
+    pub fn add_root_certificate(&mut self, cert: Certificate) -> ::Result<&mut TlsClient> {
```
Since we have a builder that needs to consume itself to create a final `Client`, we should probably follow the [consuming builder guidelines](https://github.com/brson/rust-api-guidelines#consuming-builders).

Ugh, but at the same time, there are methods that could fail, such as this one, and a `Result` is needed. That'd mean that using the consuming pattern, you could lose the builder if it errors... 

I'll have to think on this part.

little-dude replied:
This is how it's done in [`hyper-native-tls`](https://github.com/sfackler/hyper-native-tls/blob/master/src/lib.rs#L116).

I've not been very consistent here, because `danger_disable_hostname_verification` and `enable_hostname_verification` don't return anything. I should probably make them return `&mut ClientBuilder` to be consistent with `add_root_certificate`, or return `ClientBuilder` as suggested in the guildelines.

little-dude commented:
It's kind of unrelated to certificates, but I'm thinking that the `ClientBuilder` in this PR could be used to implement something similar to what @tshepang [proposed](https://github.com/seanmonstar/reqwest/issues/30#issuecomment-271712378) in the proxy issue (https://github.com/seanmonstar/reqwest/issues/30).

```rust
let client = ClientBuilder::new()?.proxy_url(url).proxy_auth(auth).build()?;
```

I'd be interested in giving it a try if this get merged, what do you think?

seanmonstar commented:
@little-dude yes, I'd expect proxy methods would be part of the `ClientBuilder`, that's why I liked it's name more than the specific `TlsClient`.

(For the actual API, I like the `Proxy` type spelled out in #30 personally, when we get there.)

little-dude commented:
For some reason, this PR is not working in my case. Certificate verification still fails.

```
Error { code: 337047686, library: "SSL routines", function: "tls_process_server_certificate", reason: "certificate verify failed", file: "ssl/statem/statem_clnt.c", line: 1245 }
```

little-dude commented:
Not sure how to debug that... both `curl` and `wget` refuse to use my der cert, but are ok with the pem version. Maybe it's because it's a bundle and I need to separate each certificate of the chain?

I'll try to find a public server with a self signed certificate to see if I have the same issue.

seanmonstar commented:
So, this doesn't quite work?

little-dude commented:
@seanmonstar I rebased, squashed, and added tests (poorly written).

I made it work yesterday with certificates I generated. But after making changed to the tests and automating certificates generation, it does not work anymore... but I'm still working on it. I'm not very familiar with TLS/openssl/certificates so I may be doing stupid mistakes.

Regarding the tests: I have a script that generates a root CA, and an intermediate CA. I use the intermediate CA to make two server certificates:

- one with the CN `localhost` that I use to test the client with a custom CA and hostname verification *enabled*.
- one with the CN `wrong.hostname.com`that I use to test the client with a custom CA and hostname verification *disabled*.

Assuming I can make these tests pass, what are your feelings about needing a shell script to run, for the tests?

little-dude commented:
I'm not sure I'll be able to debug the windows issue, so I'll close this for now, sorry...
To whoever wants to pick this up, there are two issues here:

- in 12c71fa: there seems to be a problem with schannel. See [this PR](https://github.com/sfackler/hyper-native-tls/pull/11) in `hyper-native-tls` for more details.
- in eb76046: this commit is supposed to be a harmless refactoring of the tests, but it makes tests unpredicatbly. On my local linux machine, there's roughly 50% chance that one of the tests hang. As explained in the commit message, it _might_ be due to [an open openssl issue](https://github.com/openssl/openssl/issues/3505), but this is just a wild guess.

seanmonstar commented:
I'm going to resurrect this, because there is a lot of good work in here! To start, we can just assume that rust-native-tls has tests of its own, and that we don't need to test that its own functionality does what it says. We can open a new issue for getting solid TLS tests running in reqwest on target platforms.

stephanbuys commented:
I can test on MacOS, have a bit of experience with openssl and certs in general (nothing stellar), so please ping me if I can help.

seanmonstar commented:
Re-opened in #124, without the TLS tests.

little-dude commented:
wow thanks a lot @seanmonstar. I was kind of disappointed that I could no finish this up, so I'm glad to see this was not useless. Once https://github.com/openssl/openssl/pull/3525 is merged, I'll try again to see if tests still hang on linux.

sfackler commented:
There's a new version of openssl-sys that works around the OpenSSL issue.

little-dude commented:
The tests still hang on my machine with `openssl-sys 0.9.13`.
Is there a way to debug this kind of stuff in rust? Something like `pdb` in python?

sfackler commented:
gdb and lldb work.

little-dude commented:
The hang is not directly related to openssl I think. Sometimes certificate verification fails, so the test panics, and the code that stops the server is never run:

```rust
#[test]
fn test_custom_ca_hostname_verification_disabled() {
    let listening = start_server("tests/certificates/server-wrong.hostname.com.pfx");
    let client = get_client(HostnameVerification::Disabled);
    // panic occurs here:
    // Error { kind: Http(Ssl(Failure(Ssl(ErrorStack([Error { code: 337047686, library: "SSL routines", function: "tls_process_server_certificate", reason: "certificate verify failed", file: "ssl/statem/statem_clnt.c", line: 1230 }]))))), url: Some("https://localhost:12345/") }
    let mut resp = client.get("https://localhost:12345").unwrap().send().unwrap();
    let mut body = vec![];
    resp.read_to_end(&mut body).unwrap();
    assert_eq!(body, b"ok");
    // this does not run
    mem::forget(listening);
}
```

After handling the error and calling `listening.close()` I don't see the hang anymore.
I just reopened a PR to try this against windows and mac.

The fact that _sometimes_ certificate verification fails is not good though...
